### DISCUSSION TITLE
App performance

### DISCUSSION THREAD

**Discussion by sergey-dme** [Budibase/budibase#2348] [Help] [answered]
Hey all,

Whereas BB being continuously improving and getting better and better, I am also continuously developing and improving my app, it's growing in screen number perspective, its size and functionality of course. Currently it has about 50 screens, 7 main layouts, almost empty Tables (just a few test rows/records stored), exported app file size is about 25 Mb / same size according to the Chrome' performance inspector. 

As a result, I've got quite long app loading time. When app is opening it takes **2.5 seconds** for it to be appeared/loaded, and during the loading process the only white background is showing in a browser tab, nothing else. 

I've also analyzed its performance in Chrome inspector and consider that it takes too much time to be loaded (compare to other even more complex web app and considering that my app is accessed from my local network without a long chain of different DNS or other types internet servers). In builder, app loading time is about the same.
When app is loaded it works perfectly and fast.

**I've tried to investigate it and found two main reasons leading to increasing App loading time:**

1. One of the elements of my dashboard (it's my home screen) - Device Status bar, which has 14 Data Providers inside :) - because I use them to show the status of 7 Devices (2 data providers per device) using filter and `Rows Length `with `{{#if}} `for displaying _appropriate color Device icons_ depending on the _opened issues_ and its _Priorities/Types_. Moreover, there are 13 more Data Providers in other sections of my dashboard. 
And.. all 27 Data Providers used in my dashboard are getting the data from the same Table, the only difference are Filters.
I would not deny that it is not optimized at all, but I have no other ways/solutions for this use case/task as well as the second one:

2. The UI of my app is complex but nothing unusual. I use some background wide jpg/png images (web-optimised/compressed of course), I tried to use `background-image: url("http://...") `in Custom CSS and host the images on different servers even in local network, but the bg images start blinking (known behaviour for bg images in any browser, as they should be pre-loaded that is something that is not currently possible in BB). Therefore, I ended up by using amazing base64: `background-image: url("data:image/png;base64,iVBORw0KG..` storing the images in the app screens. I know, it is usually used to handle small images, ui elements, icons etc.. but I have no choice. Thanks you for the ability to use base64 in Custom CSS anyway. Please do not remove it)

The first one (lots of data providers) has the most significant effect on app loading time. 
The second one mostly effects on app size, on the loading time - secondly in my case.

Could please advise on it, how to decrease app loading time? Thanks in advance.

`Current Version: 0.9.104`

UPD: The video showing the case:

https://user-images.githubusercontent.com/53024832/129254563-d8a2c84c-77d5-4ef9-8d31-58179698462f.mov

If I rightly understand, when app is opening - the whole app content: scripts, UI, ALL screens/pages are loaded/downloaded from the app server (as I can see in performance section of Chrome Inspector) And as for the app size, it's difficult for me to explain but I feel like there is a something unnecessary/old/already deleted still stored in my app, I don't think there is really 25MB, almost nothing heavy  there, maybe for 10-12 MB. I am unable to analyse the exported file. Sometimes I have failures when exporting/importing the app as it shown below (just an example from today):
![image](https://user-images.githubusercontent.com/53024832/129262253-79b26842-b4d7-467d-829b-ced4d8ad1802.jpeg)
But after few tries it goes away and export/import ends successfully. First time I thought the errors occurs because of the base64 in CSS and I tested it by removing all base64 images data from custom css across whole app, but it doesn't affect. Base64 works fine in BB and doesn't lead to an issue, just slightly increases the app size (depending on the image of course).

**Comment by joebudi** [comment]
@sergey-dme Thank you for the detailed write-up. As I've mentioned before, you're really pushing the boundaries of Budibase and this makes us feel quite jubilant. 

@shogunpurple is looking into this and will respond in due course.

**Comment by shogunpurple** [comment, marked as answer]
hi @sergey-dme, thanks for the detailed issue.

> I've also analyzed its performance in Chrome inspector and consider that it takes too much time to be loaded (compare to other 
> even more complex web app and considering that my app is accessed from my local network without a long chain of different 
> DNS or other types internet servers). In builder, app loading time is about the same.
> When app is loaded it works perfectly and fast.

I don't think that the issue here is network latency because it happens on every load, making the likely culprit JS execution. I have set up a fairly complex app in my environment and have found it difficult to reach anywhere close to 2.5s on initial load, but there are a lot of factors at play such as the hardware you are using to run budibase on. I ran a lighthouse test on my app and noticed that the longest script to run by far is the main budibase client library, which makes sense given it's responsible for rendering the entire app. I would imagine the situation you are seeing here is the same. Data providers load asynchronously, so the initial setup of the DOM before they even get rendered is what seems to be slow here.

I ran a few performance diagnostics in chrome as well and noticed that we are experiencing dropped frames and layout shifts when using charts (which might actually contribute to your images blinking). This should not really affect the initial loading time though. It seems like there's a lot of JS execution going in to setting up nested trees of data in the client library, especially when there are handlebars expressions there.

The best way to debug your app in depth would be to allow us to have a look at your app by exporting, but I understand if your data is private and you would rather not do that. The next best way for us to debug is to get some more diagnostics from you. It would be great if you could do the following:

- Run your app in a fresh browser with no extensions or an incognito window - this is to rule out extensions playing a part. Does the app load faster?
- Run a lighthouse test on your app in incognito (https://developers.google.com/web/tools/lighthouse/) and share the PDF result with us. This will help us to identify the slowest loading/executing scripts and how long they are actually taking
- A HAR file with your network requests on app load, which will help identify the requests being made to the server and if there is in fact any large files that are blocking initial render and not being cached (https://help.coohom.com/how-to-save-a-har-file/log-in-google-chrome)

### Import/Export
This is an interesting one in that it happens intermittedly. Is there a set of reproduction steps that you have found make it happen? I initially thought it was a timeout on the server, but it looks like it's actually having trouble parsing rows in the CSV. I'll keep digging into this one.

Just a note - the size of the import/export is not really the size of the app itself, it's just the size of the exported txt file, so it's going to be fairly large if you have a lot of data.

**Reply by sergey-dme** [reply]
Hey @shogunpurple, Thank you for attention to this, valuable explanation and offering debug my app. 
I would be happy to send the app, no private data there and I can clear some of it anyway, we just started testing my app with my colleagues. Martin, would it be okay for you if I will send the app to your bb email that I found it in your GitHub profile?

The initial loading time is the same when I'm accessing the app on server computer (localhost) using clean Chrome without any extensions installed. The server running on recently installed Win10 and has pretty standard/common hardware, but currently only 16GB RAM, I'm going to extend it to 32GB soon. The app loading time is about the same on my colleagues PCs within local network. 
Despite of that, I suspect that something wrong with my server. There are few reasons: 
1. I don't like how images look in docker - I never saw "scarf" before on my previous server: 
![scarf](https://user-images.githubusercontent.com/53024832/129354600-1291caf6-8753-4d83-a3de-eb93d3608e15.PNG)

2. Every time running `cli hosting --start `I get message: `Starting services, this may take a moment - first time this may take a few minutes to download images.` (yaml files will be attached with the app in email)
![cli](https://user-images.githubusercontent.com/53024832/129354613-97e0c63d-ccdf-4a56-bde7-a7ad45596717.PNG)

3. I'm still using cli to update the BB, as updating trough UI does not work.

You will have noticed in my app that the Device Status bar is a separate screen (with empty layout) that I call in my dashboard using Embed component and iframe. So, it might lead to loading time increasing as it like the _two apps_ are loading simultaneously, but in fact, I do this way, because it slightly **decreases** initial loading time compare to the case, when device status bar was part of the dashboard screen (it took about 3-4 sec when the section was inside the dashboard screen). One more advantage I got calling it in embed: I am able to use welcome message (as bg) for a few seconds.

**Reply by shogunpurple** [reply]
@sergey-dme 

> Martin, would it be okay for you if I will send the app to your bb email that I found it in your GitHub profile?

Yep, that would be great thanks!

> I don't like how images look in docker - I never saw "scarf" before on my previous server:

The images are the same, but we use https://scarf.sh/home as a redirect so we can get some analytics about our docker downloads.

It's odd that your UI still does not work for updating. I should be able to debug this when I see your YAML files. 

Thanks!

**Reply by sergey-dme** [reply]
Just tried to export the app, after about 20 tries, only two times I got the files, but all of them different sizes, about 17 Mb - less than it should be (tried to import both and got errors), so these files corrupted, usual size of the app is 25MB. Luckily I was able to export and send the app yesterday. I checked some logs in docker:

Redis:
```
1:C 15 Aug 2021 02:06:47.499 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:C 15 Aug 2021 02:06:47.499 # Redis version=6.2.4, bits=64, commit=00000000, modified=0, pid=1, just started
1:C 15 Aug 2021 02:06:47.499 # Configuration loaded
1:M 15 Aug 2021 02:06:47.500 * monotonic clock: POSIX clock_gettime
1:M 15 Aug 2021 02:06:47.501 * Running mode=standalone, port=6379.
1:M 15 Aug 2021 02:06:47.501 # Server initialized
1:M 15 Aug 2021 02:06:47.501 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
1:M 15 Aug 2021 02:06:47.501 * Loading RDB produced by version 6.2.4
1:M 15 Aug 2021 02:06:47.501 * RDB age 22 seconds
1:M 15 Aug 2021 02:06:47.501 * RDB memory usage when created 7.77 Mb
1:M 15 Aug 2021 02:06:47.543 * DB loaded from disk: 0.042 seconds
1:M 15 Aug 2021 02:06:47.544 * Ready to accept connections
```
UPD: Docker restart didn't help. Windows restart/Server reboot helped. App exported successfully after it.
Im thinking about the server replacement. On my office PC it worked faster and I never had any errors over the last 3 month since I started dev the app, I just moved my app to the server about a month ago and having the problems now

UPD2: Today export stopped working again. Builder says "App exported successfully" but the file does not come. Reboot does not help anymore.

UPD3: Still unable to export the app. After about 30 sec txt file comes, but ONLY in Safari, and the file size less than it should be, each try I get different txt file - different size 12-23MB. All those files are "bad", as good ones (my previous old backups/exported app) always end with `}` symbol and larger. 
In Chrome I get nothing, during the export file does not appear to download (even in clean chrome and localhost). Tried different PCs/clients and networks (localhost, local network, Internet). Results are the same. Still trying to export my app and get the file... hundreds of reboots and export tries. Many changes I made today for the app. I want to take the app out off the server, with my latest changes.

**Reply by sergey-dme** [reply]
Good morning @shogunpurple have you been able to check my app? Just want to make sure you received the email with zip attached. Thanks.

**Reply by shogunpurple** [reply]
Hi @sergey-dme - I received your email but haven't had a chance to look at it yet. I'll respond in due course

**Reply by shogunpurple** [reply]
@sergey-dme thanks a lot for your patience on this. I've pulled in your app (Which looks amazing, by the way. One of the best apps I've seen built with budibase) and debugged it in my own environment. Straight away the culprit revealed itself - the `appPackage`.

The appPackage is a large JSON object that budibase pulls from the backend, and it contains all the main building blocks of your app. Application metadata, layouts, screens, even access controls. Due to the number of screens that you have in your app, your `appPackage` is very large, coming in at around 10MB. The `appPackage` needs to be fetched before the app can render anything, hence the delay you are seeing.

This `appPackage` call is also not being cached and fetched on every initial load in the preview section. This is actually something we should optimise and look at - either by caching it into redis or treating it like a static file with a hash, allowing the browser to cache it through a service worker or similar.

Unfortunately there is not really anything that you can do at the moment until we address this - it's a matter of download speed, I'm afraid. I'm going to turn this discussion into an issue, because we should address it but I can't give you an exact timeline.

**Reply by sergey-dme** [reply]
Hi @shogunpurple Thank you very much for the kind words and your time looking into this. I'm glad that you are seeing the way to improve performance with initial load. I also appreciate this explanation about the `appPackage`.

Luckily, I solved the problem with 2.5s initial loading time by moving to a new clean server with a better hardware. Now the app takes about 1s to be loaded. Also, to reduce the app size, I replaced all base64 jpg/png by new cool base64 webp that gives better compression keeping the image quality the same and even better. Moreover, I replaced external icons (png) on my device status bar by SVGs using embed component.

Despite of my app size increased over the last week (new screens added) I don't have anymore problems with app export/import. I think this issue was happening due to my server.

The issue I still have -- it's unable to update the bb using UI (again, on a new clean server). Next week I will post the watchover log in another discussion I opened for this problem some time ago.

Thank you again.

Regards,
Sergey
